OpenAI: Function Calling (#7704)
mendix/docs: content/en/docs/appstore/modules/openai/_index.md

| `Role`        | This is the role of the message author.<br />For more information, see the [ENUM_Role](#enum-role) section. |
| `ToolCallId`  | Tool call that this message is responding to. Only applicable and required for messages with role `tool`.                |

##### 4.1.2.3 `AbstractTool` {#abstracttool}

This is the abstract entity for `Tool` reused in the Chat Completions request and response. Do not use this entity directly. Instead, use one of its specializations. 

| Attribute | Description                                                  |
| --------- | ------------------------------------------------------------ |
| `ToolType`| The type of the tool. Currently, only function is supported. <br />For more information, see the [ENUM_ToolType](#enum-tooltype) section. |

##### 4.1.2.4 `AbstractFunction` {#abstractfunction}

This is the abstract entity for `Function` reused in the Chat Completions request and response. Do not use this entity directly. Instead, use one of its specializations. 

| Attribute | Description                                                  |
| --------- | ------------------------------------------------------------ |

> | `Name`    | The name of the function to call.                            |

Reviewer: Would it be a good idea to add something here to support what you answered below:
```suggestion
| `Name`    | The name of the function to call. The model will use this to help identify which function to call where it has a choice. |
```

Reviewer 2: It also functions as UUID in other specialization so this is not correct. In my opinion the suggestion to use the description is sufficient

Reviewer: OK - I'll resolve this and not commit the suggestion.

The ToolCallFunction entity is only applicable for messages with role `assistant`.

| Attribute  | Description                                    |
| ---------- | -----------------------------------------------|
| `Arguments`| The arguments to call the function with, as generated by the model in JSON format. Note that the model does not always generate valid JSON, and may hallucinate parameters not defined by your function schema. Validate the arguments in your code before calling your function. |

##### 4.1.3.7 `Tools` {#tools}

This is a wrapper for a list of tools the model may call. Currently, only functions are supported as as tools. A maximum of 128 functions are supported.

##### 4.1.3.8 `ToolRequest` {#toolrequest}

This is a specialization of the [AbstractTool](#abstracttool) entity and represents a tool the model may call.

##### 4.1.3.9 `FunctionRequest` {#functionrequest}

This is a specialization of the [AbstractFunction](#abstractfunction) entity and represents a function the model may call.

| Attribute          | Description                                                  |
| ------------------ | ------------------------------------------------------------ |

> | `Description`      | A description of what the function does, used by the model to choose when and how to call the function. This attribute is optional. |

Reviewer: If the model is using this to choose when to call the function how can it be optional?

Reviewer 2: It's optional in the OpenAI API, so we made it optional for our connector as well. If you give the function a very descriptive name, it may work just fine without the description
https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools

Reviewer: Thanks for explaining. Should we put something into the docs to explain this?

Reviewer 2: I would phrase it in a way that we recommend adding a description even though it is optional

Reviewer: How about
```suggestion
| `Description`      | A description of what the function does, used by the model to choose when and how to call the function. This attribute is optional, but Mendix suggests using it to help the model choose the correct function. |
```

This is a specialization of the [AbstractChatCompletionsMessage](#abstractchatcompletionsmessage) entity. Each instance contains a text that needs to be taken into account by the model when processing the completion request. 

##### 4.1.3.5 `ToolCall` {#toolcall}

This is a specialization of the [AbstractTool](#abstracttool) entity and represents the tool calls generated by the model, such as function calls. The ToolCall entity is only applicable for messages with role `assistant`.

| Attribute | Description                                    |
| --------- | -----------------------------------------------|
| `_id`     | The ID of the tool call, generated by the LLM. |

##### 4.1.3.6 `ToolCallFunction` {#toolcallfunction}

This is a specialization of the [AbstractFunction](#abstractfunction) entity and represents the function that the model called.
The ToolCallFunction entity is only applicable for messages with role `assistant`.

| Attribute  | Description                                    |
| ---------- | -----------------------------------------------|

> | `Arguments`| The arguments to call the function with, as generated by the model in JSON format. Note that the model does not always generate valid JSON, and may hallucinate parameters not defined by your function schema. Validate the arguments in your code before calling your function. |

Reviewer: If the function is being called automatically you can't always validate before it is called - so making the recommendation more generic.

```suggestion
| `Arguments`| The arguments to call the function with, as generated by the model in JSON format. Note that the model does not always generate valid JSON, and may hallucinate parameters not defined by your function schema. Validate the arguments before using them. |
```

| `Configuration`  | [Configuration](#configuration-entity)                | mandatory                     | This is an object that contains endpoint and API key.        |
| `Model`          | String                                                | only mandatory for **OpenAI** | This is the ID of the model to use; not considered for **Azure OpenAI** configurations. |
| `ResponseFormat` | [ENUM_ResponseFormat_Chat](#enum-responseformat-chat) | optional                      | This can be used to specify the format that the model must output. |

> | `Temperature`    | Decimal                                               | optional                      | This can be used to control the randomness of the output. The value should be a decimal between 0.0 and 2.0. The default value is 1.0. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic. Note: high values for temperature (>1.7) may give unexpected results and even internal server errors. |

Reviewer: Since the default value is 1.0 I think it would be better to have a higher value than one as your example of a higher value. Otherwise you are saying that the default is more random than setting it explicitly to be more random.

Reviewer 2: Agree, can we just remove the "like 0.8"?

Reviewer: @tinaengel - is this OK?
```suggestion
| `Temperature`    | Decimal                                               | optional                      | This can be used to control the randomness of the output. The value should be a decimal between 0.0 and 2.0. The default value is 1.0. Higher values make the output more random, while lower values make it more focused and deterministic. Note: very high values for temperature (>1.7) may give unexpected results and even internal server errors. |
```

| `FunctionCollection`     | Object                                                | optional                      | A collection of functions to be sent along with the ChatCompletionsRequest as tools. |

**Return value**

| Name                    | Type   | Description                                                  |
| ----------------------- | ------ | ------------------------------------------------------------ |
| `AssistantResponseText` | String | Assistant message that was generated by the model as a response to a user message. |

The following microflows may be used in order to construct and handle the required inputs: 

* `ChatCompletionsSession_Create` is used to create the session wrapper that must be passed as input parameter. 

> * `ChatCompletionsSession_AddMessage` is used to attach the historical messages to the `ChatCompletionsSession`. The content of such a message corresponds to a system, assistant, or user prompt. In the case of multiple historical messages the order is relevant.

Reviewer: In what way relevant? Which one takes priority, first or last?

Reviewer 2: it's a chronological order, from oldest message to most recent message being the last one

Author: It's not really about priority, but rather the chronological order. So your last messages are the most current, while the first ones are "older". The model responds to the last message, but takes into account the older ones (like a normal conversation).

Reviewer: Would it be worth making this explicit?
```suggestion
* `ChatCompletionsSession_AddMessage` is used to attach the historical messages to the `ChatCompletionsSession`. The content of such a message corresponds to a system, assistant, or user prompt. In the case of multiple historical messages the latest message has more influence than earlier ones.
```

Reviewer 2: How about: "If multiple messages are present, they should be ordered chronologically."

Reviewer: Great suggestion - I'll put it here and commit it.

```suggestion
* `ChatCompletionsSession_AddMessage` is used to attach the historical messages to the `ChatCompletionsSession`. The content of such a message corresponds to a system, assistant, or user prompt. If multiple messages are present, they should be ordered chronologically.
```

Reviewer: Wrong suggestion committed. Now committed the right suggestion 54022f4b47c43a6cfe084a809db1977edd88fae5

| `Configuration`          | [Configuration](#configuration-entity)                | mandatory                     | This is an object that contains endpoint and API key.        |
| `Model`                  | String                                                | only mandatory for **OpenAI** | This is the ID of the model to use; not considered for **Azure OpenAI** configurations. |
| `ResponseFormat`         | [ENUM_ResponseFormat_Chat](#enum-responseformat-chat) | optional                      | This can be used to specify the format that the model must output. |

> | `Temperature`            | Decimal                                               | optional                      | This can be used to control the randomness of the output. The value should be a decimal between 0.0 and 2.0. The default value is 1.0. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic. Note: high values for temperature (>1.7) may give unexpected results and even internal server errors. |

Reviewer: ```suggestion
| `Temperature` | Decimal | optional | This can be used to control the randomness of the output. The value should be a decimal between 0.0 and 2.0. The default value is 1.0. Higher values make the output more random, while lower values make it more focused and deterministic. Note: very high values for temperature (>1.7) may give unexpected results and even internal server errors. |
```

| `FunctionCollection`     | Object                                                | optional                      | A collection of functions to be sent along with the ChatCompletionsRequest as tools. |

**Return value**

| Name                    | Type   | Description                                                  |
| ----------------------- | ------ | ------------------------------------------------------------ |
| `AssistantResponseText` | String | Assistant message that was generated by the model as a response to a user message. |

The following microflows may be used to construct and handle the required inputs: 

* `ChatCompletionsSession_Create` is used to create the session wrapper that must be passed as input parameter. 

> * `ChatCompletionsSession_AddMessage` is used to attach the historical messages to the `ChatCompletionsSession`. The content of such a message corresponds to a system, assistant, or user prompt. In the case of multiple historical messages the latest message has more influence than earlier ones.

Reviewer: ```suggestion
* `ChatCompletionsSession_AddMessage` is used to attach the historical messages to the `ChatCompletionsSession`. The content of such a message corresponds to a system, assistant, or user prompt. If multiple messages are present, they should be ordered chronologically.
```
